// mapmyindia-search-widgets-react-native/index.js
import MapmyIndiaPlacePicker from './src/MapmyIndiaPlacePicker';
import Components, {PlaceWidgetProps} from './src/components';

import {NativeModules} from 'react-native';
const {MapmyindiaReactNativePlacePicker} = NativeModules;
const {
  POD_SUB_LOCALITY_KEY,
  POD_LOCALITY_KEY,
  POD_CITY_KEY,
  POD_VILLAGE_KEY,
  POD_SUB_DISTRICT_KEY,
  POD_DISTRICT_KEY,
  POD_STATE_KEY,
  POD_SUB_SUB_LOCALITY_KEY,
  GRAVITY_LEFT_KEY,
  GRAVITY_CENTER_KEY,
  GRAVITY_RIGHT_KEY,
  GRAVITY_TOP_KEY,
  GRAVITY_BOTTOM_KEY,
  SIZE_SMALL_KEY,
  SIZE_MEDIUM_KEY,
  SIZE_LARGE_KEY,
} = MapmyindiaReactNativePlacePicker.getConstants();
const MapmyIndiaUIWidgets = {
  searchWidget: async (props: PlaceWidgetProps) =>
    Components.openPlaceWidget(props),
  PlacePicker: MapmyIndiaPlacePicker,
  //Pods
  POD_SUB_LOCALITY: POD_SUB_LOCALITY_KEY,
  POD_LOCALITY: POD_LOCALITY_KEY,
  POD_CITY: POD_CITY_KEY,
  POD_VILLAGE: POD_VILLAGE_KEY,
  POD_SUB_DISTRICT: POD_SUB_DISTRICT_KEY,
  POD_DISTRICT: POD_DISTRICT_KEY,
  POD_STATE: POD_STATE_KEY,
  POD_SUB_SUB_LOCALITY: POD_SUB_SUB_LOCALITY_KEY,

  //attributionHorizontalAlignment
  GRAVITY_LEFT: GRAVITY_LEFT_KEY,
  GRAVITY_CENTER: GRAVITY_CENTER_KEY,
  GRAVITY_RIGHT: GRAVITY_RIGHT_KEY,

  //attributionVerticalAlignment
  GRAVITY_TOP: GRAVITY_TOP_KEY,
  GRAVITY_BOTTOM: GRAVITY_BOTTOM_KEY,

  //logoSize
  SIZE_SMALL: SIZE_SMALL_KEY,
  SIZE_MEDIUM: SIZE_MEDIUM_KEY,
  SIZE_LARGE: SIZE_LARGE_KEY,
};

export default MapmyIndiaUIWidgets;
