// mapmyindia-map-react-native-beta/javascript/modules/location/locationManager.js
import {NativeModules, NativeEventEmitter} from 'react-native';

const MapmyIndiaGL = NativeModules.MGLModule;
const MapmyIndiaGLLocationManager = NativeModules.MGLLocationModule;

export const LocationModuleEventEmitter = new NativeEventEmitter(
  MapmyIndiaGLLocationManager,
);

class LocationManager {
  constructor() {
    this._listeners = [];
    this._lastKnownLocation = null;
    this._isListening = false;
    this._isPaused = false;
    this.onUpdate = this.onUpdate.bind(this);
  }

  async getLastKnownLocation() {
    if (!this._lastKnownLocation) {
      const lastKnownLocation = await MapmyIndiaGLLocationManager.getLastKnownLocation();
      if (!this._lastKnownLocation && lastKnownLocation) {
        this._lastKnownLocation = lastKnownLocation;
      }
    }
    return this._lastKnownLocation;
  }

  addListener(listener) {
    if (!this._listeners.includes(listener)) {
      this._listeners.push(listener);

      if (this._lastKnownLocation) {
        listener(this._lastKnownLocation);
      }
    }
  }

  removeListener(listener) {
    this._listeners = this._listeners.filter(l => l !== listener);
  }

  removeAllListeners() {
    this._listeners = [];
  }

  start(displacement = 0) {
    if (this._isPaused) {
      MapmyIndiaGLLocationManager.start(displacement);
      this._isPaused = false;
      return;
    }

    if (!this._isListening) {
      MapmyIndiaGLLocationManager.start(displacement);

      LocationModuleEventEmitter.addListener(
        MapmyIndiaGL.LocationCallbackName.Update,
        this.onUpdate,
      );

      this._isListening = true;
    }
  }

  pause() {
    if (!this._isPaused && this._isListening) {
      MapmyIndiaGLLocationManager.pause();
      this._isListening = false;
    }
  }

  dispose() {
    MapmyIndiaGLLocationManager.stop();

    if (this._isListening) {
      LocationModuleEventEmitter.removeListener(
        MapmyIndiaGL.LocationCallbackName.Update,
        this.onUpdate,
      );
    }

    this._isListening = false;
  }

  setMinDisplacement(minDisplacement) {
    MapmyIndiaGLLocationManager.setMinDisplacement(minDisplacement);
  }

  onUpdate(location) {
    this._lastKnownLocation = location;

    for (const listener of this._listeners) {
      listener(location);
    }
  }
}

export default new LocationManager();
