// mapmyindia-map-react-native-beta/javascript/modules/restApi/RestApi.js
 import { number } from 'prop-types';
import { Platform } from 'react-native';
import {NativeModules, NativeEventEmitter} from 'react-native';

const Api = NativeModules.MGLRestApiModule; 

const RestApi = {

    autoSuggest: async function(params) {
       return  Api.autoSuggest(params) 
    },

    reverseGeocode: async function(params) {
        return  Api.reverseGeocode(params) 
    },

    geocode: async function(params) {
        return  Api.geocode(params) 
    },

    textSearch: async function(params) {
       return Api.textSearch(params) 
    },

    placeDetail: async function(params) {
       return  Api.placeDetail(params) 
    },
    
    placeDetailLegacy :async function(params) {
       return Api.placeDetailLegacy(params) 
    },
    nearby: async function(params) {
        return  Api.nearby(params) 
    },

    direction: async function(params) {
       return  Api.direction(params)
    },

    distance: async function(params) {
        return   Api.distance(params)
    },
    POIAlongRoute : async function(params) {
       return Api.POIAlongRoute(params)
    },
    feedback: async function(params) {
         return  Api.feedback(params)
    },
    
    mapmyIndiaClusterLinkedDevices: async function(params) {
       return Api.MapmyIndiaClusterLinkedDevices(params)
 },

   mapmyIndiaDeleteClusterLinkedDevice: async function(params) {
   
    return Api.MapmyIndiaDeleteClusterLinkedDevice(params)
},
  
   mapmyIndiaEndSession: async function(params){
       
    return Api.MapmyIndiaEndSession(params);
   }

};
RestApi.AutoSuggestCriteria = {};
RestApi.AutoSuggestCriteria.POD_CITY = Api.AutoSuggestCriteria.POD_CITY;
RestApi.AutoSuggestCriteria.POD_DISTRICT = Api.AutoSuggestCriteria.POD_DISTRICT;
RestApi.AutoSuggestCriteria.POD_LOCALITY = Api.AutoSuggestCriteria.POD_LOCALITY;
RestApi.AutoSuggestCriteria.POD_STATE = Api.AutoSuggestCriteria.POD_STATE;
RestApi.AutoSuggestCriteria.POD_SUB_DISTRICT = Api.AutoSuggestCriteria.POD_SUB_DISTRICT;
RestApi.AutoSuggestCriteria.POD_SUB_LOCALITY = Api.AutoSuggestCriteria.POD_SUB_LOCALITY;
RestApi.AutoSuggestCriteria.POD_SUB_SUB_LOCALITY = Api.AutoSuggestCriteria.POD_SUB_SUB_LOCALITY;
RestApi.AutoSuggestCriteria.POD_VILLAGE = Api.AutoSuggestCriteria.POD_VILLAGE;


//GeoCoding Constants only for ANDORID
RestApi.GeoCodingCriteria = {};
RestApi.GeoCodingCriteria.BIAS_DEFAULT = Api.GeoCodingCriteria.BIAS_DEFAULT;
RestApi.GeoCodingCriteria.BIAS_RURAL = Api.GeoCodingCriteria.BIAS_RURAL;
RestApi.GeoCodingCriteria.BIAS_URBAN = Api.GeoCodingCriteria.BIAS_URBAN;

RestApi.GeoCodingCriteria.POD_CITY = Api.GeoCodingCriteria.POD_CITY;
RestApi.GeoCodingCriteria.POD_DISTRICT = Api.GeoCodingCriteria.POD_DISTRICT;
RestApi.GeoCodingCriteria.POD_HOUSE_NAME = Api.GeoCodingCriteria.POD_HOUSE_NAME;
RestApi.GeoCodingCriteria.POD_HOUSE_NUMBER = Api.GeoCodingCriteria.POD_HOUSE_NUMBER;
RestApi.GeoCodingCriteria.POD_LOCALITY = Api.GeoCodingCriteria.POD_LOCALITY;
RestApi.GeoCodingCriteria.POD_PINCODE = Api.GeoCodingCriteria.POD_PINCODE;
RestApi.GeoCodingCriteria.POD_POINT_OF_INTEREST = Api.GeoCodingCriteria.POD_POINT_OF_INTEREST;
RestApi.GeoCodingCriteria.POD_STATE = Api.GeoCodingCriteria.POD_STATE;
RestApi.GeoCodingCriteria.POD_STREET = Api.GeoCodingCriteria.POD_STREET;
RestApi.GeoCodingCriteria.POD_SUB_DISTRICT = Api.GeoCodingCriteria.POD_SUB_DISTRICT;
RestApi.GeoCodingCriteria.POD_SUB_LOCALITY = Api.GeoCodingCriteria.POD_SUB_LOCALITY;
RestApi.GeoCodingCriteria.POD_SUB_SUB_LOCALITY = Api.GeoCodingCriteria.POD_SUB_SUB_LOCALITY;
RestApi.GeoCodingCriteria.POD_VILLAGE = Api.GeoCodingCriteria.POD_VILLAGE;

RestApi.NearbyCriteria ={};
RestApi.NearbyCriteria.DISTANCE_ASCENDING = Api.NearbyCriteria.DISTANCE_ASCENDING;
RestApi.NearbyCriteria.DISTANCE_DESCENDING = Api.NearbyCriteria.DISTANCE_DESCENDING;

//Name options only for ANDROID
RestApi.NearbyCriteria.NAME_ASCENDING = Api.NearbyCriteria.NAME_ASCENDING;
RestApi.NearbyCriteria.NAME_DESCENDING = Api.NearbyCriteria.NAME_DESCENDING;

RestApi.NearbyCriteria.DISTANCE = Api.NearbyCriteria.DISTANCE;
RestApi.NearbyCriteria.IMPORTANCE = Api.NearbyCriteria.IMPORTANCE;

 RestApi.DirectionsCriteria ={};
// RestApi.DirectionsCriteria.EXCLUDE_TUNNEL = Api.DirectionsCriteria.EXCLUDE_TUNNEL;
// RestApi.DirectionsCriteria.EXCLUDE_RESTRICTED = Api.DirectionsCriteria.EXCLUDE_RESTRICTED;
// RestApi.DirectionsCriteria.EXCLUDE_MOTORWAY = Api.DirectionsCriteria.EXCLUDE_MOTORWAY;
// RestApi.DirectionsCriteria.EXCLUDE_FERRY = Api.DirectionsCriteria.EXCLUDE_FERRY;
// RestApi.DirectionsCriteria.EXCLUDE_TOLL = Api.DirectionsCriteria.EXCLUDE_TOLL;
 
// RestApi.DirectionsCriteria.ANNOTATION_SPEED = Api.DirectionsCriteria.ANNOTATION_SPEED;
// RestApi.DirectionsCriteria.ANNOTATION_NODES = Api.DirectionsCriteria.ANNOTATION_NODES;
// RestApi.DirectionsCriteria.ANNOTATION_MAXSPEED = Api.DirectionsCriteria.ANNOTATION_MAXSPEED;
// RestApi.DirectionsCriteria.ANNOTATION_DURATION = Api.DirectionsCriteria.ANNOTATION_DURATION;
// RestApi.DirectionsCriteria.ANNOTATION_DISTANCE = Api.DirectionsCriteria.ANNOTATION_DISTANCE;
// RestApi.DirectionsCriteria.ANNOTATION_BASE_DURATION = Api.DirectionsCriteria.ANNOTATION_BASE_DURATION;
// RestApi.DirectionsCriteria.ANNOTATION_CONGESTION = Api.DirectionsCriteria.ANNOTATION_CONGESTION;
// RestApi.DirectionsCriteria.ANNOTATION_SPEED_LIMIT = Api.DirectionsCriteria.ANNOTATION_SPEED_LIMIT;

RestApi.DirectionsCriteria.OVERVIEW_FALSE = Api.DirectionsCriteria.OVERVIEW_FALSE;
RestApi.DirectionsCriteria.OVERVIEW_SIMPLIFIED = Api.DirectionsCriteria.OVERVIEW_SIMPLIFIED;
RestApi.DirectionsCriteria.OVERVIEW_FULL = Api.DirectionsCriteria.OVERVIEW_FULL;

RestApi.DirectionsCriteria.PROFILE_WALKING = Api.DirectionsCriteria.PROFILE_WALKING;
RestApi.DirectionsCriteria.PROFILE_TRUCKING = Api.DirectionsCriteria.PROFILE_TRUCKING;
RestApi.DirectionsCriteria.PROFILE_BIKING = Api.DirectionsCriteria.PROFILE_BIKING;
RestApi.DirectionsCriteria.PROFILE_DRIVING = Api.DirectionsCriteria.PROFILE_DRIVING;

RestApi.DirectionsCriteria.RESOURCE_ROUTE_TRAFFIC = Api.DirectionsCriteria.RESOURCE_ROUTE_TRAFFIC;
RestApi.DirectionsCriteria.RESOURCE_ROUTE = Api.DirectionsCriteria.RESOURCE_ROUTE;
RestApi.DirectionsCriteria.RESOURCE_ROUTE_ETA = Api.DirectionsCriteria.RESOURCE_ROUTE_ETA;
RestApi.DirectionsCriteria.RESOURCE_DISTANCE = Api.DirectionsCriteria.RESOURCE_DISTANCE;
RestApi.DirectionsCriteria.RESOURCE_DISTANCE_ETA = Api.DirectionsCriteria.RESOURCE_DISTANCE_ETA;
RestApi.DirectionsCriteria.RESOURCE_DISTANCE_TRAFFIC = Api.DirectionsCriteria.RESOURCE_DISTANCE_TRAFFIC;

// RestApi.DirectionsCriteria.APPROACH_CURB = Api.DirectionsCriteria.APPROACH_CURB;
// RestApi.DirectionsCriteria.APPROACH_UNRESTRICTED = Api.DirectionsCriteria.APPROACH_UNRESTRICTED;

// RestApi.DirectionsCriteria.DESTINATION_ANY = Api.DirectionsCriteria.DESTINATION_ANY;
// RestApi.DirectionsCriteria.DESTINATION_LAST = Api.DirectionsCriteria.DESTINATION_LAST;

RestApi.DirectionsCriteria.GEOMETRY_POLYLINE = Api.DirectionsCriteria.GEOMETRY_POLYLINE;
RestApi.DirectionsCriteria.GEOMETRY_POLYLINE6 = Api.DirectionsCriteria.GEOMETRY_POLYLINE6;

RestApi.DirectionsCriteria.IMPERIAL = Api.DirectionsCriteria.IMPERIAL;
RestApi.DirectionsCriteria.METRIC = Api.DirectionsCriteria.METRIC;

// RestApi.DirectionsCriteria.SOURCE_ANY = Api.DirectionsCriteria.SOURCE_ANY;
// RestApi.DirectionsCriteria.SOURCE_FIRST = Api.DirectionsCriteria.SOURCE_FIRST;

RestApi.POICriteria ={};

RestApi.POICriteria.GEOMETRY_BASE64 = Api.POICriteria.GEOMETRY_BASE64;
RestApi.POICriteria.GEOMETRY_POLYLINE5 = Api.POICriteria.GEOMETRY_POLYLINE5;
RestApi.POICriteria.GEOMETRY_POLYLINE6 = Api.POICriteria.GEOMETRY_POLYLINE6;

RestApi.SessionCriteria ={};
RestApi.SessionCriteria.SESSION_TYPE_GLOBAL = Api.SessionCriteria.SESSION_TYPE_GLOBAL;
RestApi.SessionCriteria.SESSION_TYPE_NAVIGATION = Api.SessionCriteria.SESSION_TYPE_NAVIGATION;
export default RestApi
